beacon-chain/db/filesystem/data_column.go

```diff
@@ -0,0 +1,1012 @@
+		pruneMu         sync.RWMutex
+	}
+
+	// DataColumnStorageOption is a functional option for configuring a DataColumnStorage.
+	DataColumnStorageOption func(*DataColumnStorage) error
+
+	// DataColumnIdent is the unique identifier for a data column sidecar.
+	DataColumnsIdent struct {
+		Root    [fieldparams.RootLength]byte
+		Epoch   primitives.Epoch
+		Indices []uint64
+	}
+
+	storageIndices struct {
+		indices [mandatoryNumberOfColumns]byte
+		count   int64
+	}
+
+	metadata struct {
+		indices                         *storageIndices
+		sszEncodedDataColumnSidecarSize uint32
+		fileSize                        int64
+	}
+
+	fileMetadata struct {
+		period    uint64
+		epoch     primitives.Epoch
+		blockRoot [fieldparams.RootLength]byte
+	}
+
+	muChan struct {
+		mu      *sync.RWMutex
+		toStore chan []blocks.VerifiedRODataColumn
+	}
+)
+
+// WithDataColumnBasePath is a required option that sets the base path of data column storage.
+func WithDataColumnBasePath(base string) DataColumnStorageOption {
+	return func(b *DataColumnStorage) error {
+		b.base = base
+		return nil
+	}
+}
+
+// WithDataColumnRetentionEpochs is an option that changes the number of epochs data columns will be persisted.
+func WithDataColumnRetentionEpochs(e primitives.Epoch) DataColumnStorageOption {
+	return func(b *DataColumnStorage) error {
+		b.retentionEpochs = e
+		return nil
+	}
+}
+
+// WithDataColumnFs allows the afero.Fs implementation to be customized.
+// Used by tests to substitute an in-memory filesystem.
+func WithDataColumnFs(fs afero.Fs) DataColumnStorageOption {
+	return func(b *DataColumnStorage) error {
+		b.fs = fs
+		return nil
+	}
+}
+
+// NewDataColumnStorage creates a new instance of the DataColumnStorage object. Note that the implementation of DataColumnStorage may
+// attempt to hold a file lock to guarantee exclusive control of the data column storage directory, so this should only be
+// initialized once per beacon node.
+func NewDataColumnStorage(ctx context.Context, opts ...DataColumnStorageOption) (*DataColumnStorage, error) {
+	storage := &DataColumnStorage{
+		DataColumnFeed: new(event.Feed),
+		muChans:        make(map[[fieldparams.RootLength]byte]*muChan),
+	}
+
+	for _, o := range opts {
+		if err := o(storage); err != nil {
+			return nil, errors.Wrap(err, "failed to create data column storage")
+		}
+	}
+
+	// Allow tests to set up a different fs using WithFs.
+	if storage.fs == nil {
+		if storage.base == "" {
+			return nil, errNoBasePath
+		}
+
+		storage.base = path.Clean(storage.base)
+		if err := file.MkdirAll(storage.base); err != nil {
+			return nil, errors.Wrapf(err, "failed to create data column storage at %s", storage.base)
+		}
+
+		storage.fs = afero.NewBasePathFs(afero.NewOsFs(), storage.base)
+	}
+
+	storage.cache = newDataColumnStorageSummaryCache()
+
+	async.RunEvery(ctx, prunePeriod, func() {
+		storage.pruneMu.Lock()
+		defer storage.pruneMu.Unlock()
+
+		storage.prune()
```

Comment:
`prune` has its own lock. Why do we have another layer of mutex here?

Reply:
I added this commit to makes the code more clear: [0c5a12a](https://github.com/OffchainLabs/prysm/pull/15257/commits/0c5a12a579ff3dc7f4f9e9cb58523938919ce585).

`dcs.mu` is only here to protect `dcs.muChans` (as stated in 

```go
mu      sync.Mutex // protects muChans
```
)

We need the additionnal `pruneMu` to ensure no operation is in running (reading/writing etc...) when the pruning operation is in progress. We can see the pruning operation like a garbage collector. We don't want other operations to interfere with it.